src/context/editorReducers.js

```diff
@@ -13,33 +13,97 @@
 export const UPDATE_OFFLINE_TD = 'UPDATE_OFFLINE_TD';
 export const UPDATE_IS_MODFIED = 'UPDATE_IS_MODFIED';
 export const SET_FILE_HANDLE = 'SET_FILE_HANDLE';
+export const REMOVE_FORM_FROM_TD = 'REMOVE_FORM_FROM_TD';
+export const ADD_FORM_FROM_TD = 'ADD_FORM_FROM_TD';
+export const ADD_ACTIONFORM_TO_TD = 'ADD_ACTIONFORM_TO_TD';
+export const ADD_EVENTFORM_TO_TD = 'ADD_EVENTFORM_TO_TD';
 
 const updateOfflineTDReducer = (offlineTD, state) => {
-    console.log('updateofflineTD')
-    return { ...state, offlineTD: offlineTD, isModified: true};
+  return { ...state, offlineTD, isModified: true };
 };
 
+const removeFormReducer = (form, state) => {
+  let offlineTD = JSON.parse(state.offlineTD)
+  console.log(form);
+  try {
+    offlineTD[form.type][form.propName].forms.forEach((element, i) => {
+      if (typeof (element.op) === 'string') {
+        if (element.href === form.form.href) {
+          offlineTD[form.type][form.propName].forms.splice(i, 1)
+        }
+      } else {
+        if (element.href === form.form.href && element.op.indexOf(form.form.op) !== -1) {
+          element.op.splice(element.op.indexOf(form.form.op), 1)
+        }
+        if (element.op.length === 0) {
+          offlineTD[form.type][form.propName].forms.splice(i, 1)
+        }
+      }
+    });
+  } catch (e) {
+    alert('Sorry we were unable to delete the Form, please try again later');
```

Comment:
I don't think that is a useful alert.. Ifguess if a user can't delete a Form he should try again immediately or delete it manually in the JSON editor.. Just waiting and trying it again later should not change the outcome, should it?

Reply:
you are right 😄